std/assembly/util/sort.ts

```diff
@@ -43,100 +56,241 @@ export function COMPARATOR<T>(): (a: T, b: T) => i32 {
   }
 }
 
-// @ts-ignore: decorator
-@inline
+// Power Sort implementation (stable) from paper "Nearly-Optimal Mergesorts"
+// https://arxiv.org/pdf/1805.04154.pdf
+// This method usually outperform TimSort.
 export function SORT<T>(
   dataStart: usize,
   length: i32,
   comparator: (a: T, b: T) => i32
 ): void {
-  if (isReference<T>()) {
-    // TODO replace this to faster stable sort (TimSort) when it implemented
-    insertionSort<T>(dataStart, length, comparator);
-  } else {
-    if (length < 256) {
-      insertionSort<T>(dataStart, length, comparator);
-    } else {
-      weakHeapSort<T>(dataStart, length, comparator);
+  var hi = length - 1, n = length;
+  if (n <= INSERTION_SORT_THRESHOLD) {
+    if (ASC_SHRINK_LEVEL < 1) {
+      switch (n) {
+        case 3: {
+          let a = load<T>(dataStart, 1 << alignof<T>());
+          let b = load<T>(dataStart, 2 << alignof<T>());
+          let c = comparator(a, b);
+          store<T>(dataStart, select<T>(a, b, c >>> 31), 1 << alignof<T>());
+          b = select<T>(b, a, c >>> 31);
+          a = load<T>(dataStart, 0);
+          c = comparator(a, b);
+          store<T>(dataStart, select<T>(a, b, c >>> 31), 0);
+          store<T>(dataStart, select<T>(b, a, c >>> 31), 2 << alignof<T>());
```

Comment:
```suggestion
          let c = comparator(a, b) < 0;
          store<T>(dataStart, select<T>(a, b, c), 1 << alignof<T>());
          b = select<T>(b, a, c);
          a = load<T>(dataStart, 0);
          c = comparator(a, b) < 0;
          store<T>(dataStart, select<T>(a, b, c), 0);
          store<T>(dataStart, select<T>(b, a, c), 2 << alignof<T>());
```

Reply:
Originally it was `x < 0` but I figured out `x >>> 31` approx 5-7% faster

Reply:
Hmm, 5-7% seems too much for something like that. What happens with `comparator(a, b) >>> 31`?

Reply:
I guess we could add `TODO` about this and refactor to `x < 0` in the future when binaryen will support this peephole optimization. Unfortunately my previous attempts [was fail](https://github.com/WebAssembly/binaryen/pull/2868)

Reply:
So, what if just the assignment of `c` is changed to `comparator(a, b) >>> 31`? That'd replace four shifts with just one at least.